docs/fsharp/style-guide/conventions.md

```diff
@@ -0,0 +1,475 @@
+
+However, use of `mutable` in F# may feel at odds with functional purity. This is actually fine, if you adjust expectations from purity to referential transparency. We argue that referential transparency - not purity - is the end goal when writing F# functions. This allows you to write a functional interface over a mutation-based implementation for performance critical code.
+
+### Wrap mutatable code in immutable interfaces
+
+With referential transparency as a goal, it is critical to write code which does not expose the mutable underbelly of performance-critical functions. For example, the following code implements the `Array.contains` function in the F# core library:
+
+```fsharp
+[<CompiledName("Contains")>]
+let inline contains value (array:'T[]) =
+    checkNonNull "array" array
+    let mutable state = false
+    let mutable i = 0
+    while not state && i < array.Length do
+        state <- value = array.[i]
+        i <- i + 1
+    state
+```
+
+Calling this function multiple times does not change the underlying array, nor does it require you to maintain any mutable state by yourself. It is referentially transparent, even though almost every line of code inside of it uses mutation.
+
+### Consider encapsulating mutable data in classes
+
+The previous example used a single function to encapsulate operations using mutable data. This is not always sufficient for more complicated information. Consider the following sets of functions:
+
+```fsharp
+let addToClosureTable (t: Dictionary<_,_>) (a, b) =
+    if not (t.Contains(a)) then
+        t.[1] <- HashSet<_>(HashIdentity.Structural)
+    
+    t.[a].Add(b)
+
+let closureTableCount (t: Dictionary<_,_>) = t.Count
+
+let closureTableContains (t: Dictionary<_, HashSet<_>>) (a, b) =
+    t.ContainsKey(a) && t.[a].Contains(b)
+```
+
+This code is performant, but it exposes the mutation-based data structure which callers are responsible for maintaining. This can be wrapped inside of a class with no underlying members that can change:
+
+```fsharp
+/// The results of computing the LALR(1) closure of an LR(0) kernel
+type Closure1Table() =
+    let t = Dictionary<Item0, HashSet<TerminalIndex>>()
+
+    member __.Add(a, b) =
+        if not (t.Contains(a)) then
+            t.[1] <- HashSet<_>(HashIdentity.Structural)
+    
+        t.[a].Add(b)
+
+    member __.Count = t.Count
+
+    member __.Contains(a, b) = t.ContainsKey(a) && t.[a].Contains(b)
+```
+
+Because this class has no members which can change, and its binding is immutable, it is also effectively immutable. Additionally, it safely encapsulates the underlying mutation-based data structure. Classes are a powerful way to encapsulate data and routines which are mutation-based without exposing the details to callers.
+
+### Prefer `let mutable` to reference cells
+
+Reference cells can be a powerful mechanism for pointer arithemetic, but are generally not ideal for other performance-critical code. Consider the following function:
```

Comment:
arithemetic. Also I have no idea why they can be "can be a powerful mechanism for pointer arithemetic" ?